src/push_service/cdn.rs

```diff
@@ -156,4 +342,98 @@ impl PushService {
 
         Ok(())
     }
+
+    async fn upload_to_cdn2(
+        &mut self,
+        resumable_url: &Url,
+        content_length: u64,
+        mut content: impl std::io::Read + std::io::Seek + Send,
+    ) -> Result<AttachmentDigest, ServiceError> {
+        let resume_info = self
+            .get_attachment_resume_info_cdn2(resumable_url, content_length)
+            .await?;
+
+        let mut digester =
+            crate::digeststream::DigestingReader::new(&mut content);
+
+        let mut buf = Vec::new();
+        digester.read_to_end(&mut buf)?;
+
+        trace!("digested content");
+
+        let mut request = self.request(
+            Method::PUT,
+            Endpoint::cdn_url(2, resumable_url),
+            HttpAuthOverride::Unidentified,
+        )?;
+
+        if let Some(content_range) = resume_info.content_range {
+            request = request.header(CONTENT_RANGE, content_range);
+        }
+
+        request.body(buf).send().await?.error_for_status()?;
+
+        Ok(AttachmentDigest {
+            digest: digester.finalize(),
+            incremental_digest: None,
+            incremental_mac_chunk_size: 0,
+        })
+    }
+
+    async fn upload_to_cdn3(
+        &mut self,
+        resumable_url: &Url,
+        headers: &HashMap<String, String>,
+        content_type: &str,
+        content_length: u64,
+        mut content: impl std::io::Read + std::io::Seek + Send,
+    ) -> Result<AttachmentDigest, ServiceError> {
```

Comment:
I would add tracing::instrument here, and ensure that the parameters are quite simple. We have some tracing events in the body.

Reply:
Just forgot 😄 will do.